community/values/src/main/java/org/neo4j/values/storable/PointValue.java

```diff
@@ -255,4 +261,268 @@ public boolean withinRange( PointValue lower, boolean includeLower, PointValue u
+                crs = coordinates.length == 3 ? CoordinateReferenceSystem.Cartesian_3D : CoordinateReferenceSystem.Cartesian;
+            }
+        }
+        else if ( latitudeValue != null && longitudeValue != null )
+        {
+            double x = ((NumberValue) longitudeValue).doubleValue();
+            double y = ((NumberValue) latitudeValue).doubleValue();
+            AnyValue zValue = fields[PointValueField.Z.ordinal()];
+            AnyValue heightValue = fields[PointValueField.HEIGHT.ordinal()];
+            if ( zValue != null )
+            {
+                coordinates = new double[]{x, y, ((NumberValue) zValue).doubleValue()};
+            }
+            else if ( heightValue != null )
+            {
+                coordinates = new double[]{x, y, ((NumberValue) heightValue).doubleValue()};
+            }
+            else
+            {
+                coordinates = new double[]{x, y};
+            }
+            if ( crs == null )
+            {
+                crs = coordinates.length == 3 ? CoordinateReferenceSystem.WGS84_3D : CoordinateReferenceSystem.WGS84;
+            }
+            if ( !crs.isGeographic() )
+            {
+                throw new IllegalArgumentException( "Geographic points does not support coordinate reference system: " + crs );
+            }
+        }
+        else
+        {
+            if ( crs == CoordinateReferenceSystem.Cartesian )
+            {
+                throw new IllegalArgumentException( "A " + CoordinateReferenceSystem.Cartesian.getName() + " point must contain 'x' and 'y'" );
+            }
+            else if ( crs == CoordinateReferenceSystem.Cartesian_3D )
+            {
+                throw new IllegalArgumentException( "A " + CoordinateReferenceSystem.Cartesian_3D.getName() + " point must contain 'x', 'y' and 'z'" );
+            }
+            else if ( crs == CoordinateReferenceSystem.WGS84 )
+            {
+                throw new IllegalArgumentException( "A " + CoordinateReferenceSystem.WGS84.getName() + " point must contain 'latitude' and 'longitude'" );
+            }
+            else if ( crs == CoordinateReferenceSystem.WGS84_3D )
+            {
+                throw new IllegalArgumentException( "A " + CoordinateReferenceSystem.WGS84_3D.getName() +
+                                                    " point must contain 'latitude', 'longitude' and 'height'" );
+            }
+            throw new IllegalArgumentException( "A point must contain either 'x' and 'y' or 'latitude' and 'longitude'" );
+        }
+
+        if ( crs.getDimension() != coordinates.length )
+        {
+            throw new IllegalArgumentException( "Cannot create " + crs.getDimension() + "D point with " + coordinates.length + " coordinates" );
+        }
+        return Values.pointValue( crs, coordinates );
+    }
+
+    private enum PointValueField
+    {
+        X
+                {
+                    @Override
+                    ValueGroup valueType()
```

Comment:
This becomes way shorter and more readable when you create a field in this enum and a Constructor, instead of overriding the same method everywhere.

E.g.:

```
 private enum PointValueField
    {
        X( NUMBER );

        private ValueGroup valType;

        PointValueField( ValueGroup valType )
        {
            this.valType = valType;
        }

        ValueGroup valueType()
        {
            return ValueGroup.NO_VALUE;
        }
    }
```